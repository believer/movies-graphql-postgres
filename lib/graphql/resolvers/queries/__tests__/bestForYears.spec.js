const bestForYears = require('../bestForYears')
const { verifyToken } = require('../../../../services/token')

jest.mock('../../../../services/token', () => ({
  verifyToken: jest.fn().mockImplementation(() =>
    Promise.resolve({
      id: 2,
    })
  ),
}))

describe('queries/bestForYears', () => {
  let db

  beforeEach(() => {
    db = {
      query: jest.fn().mockReturnValue('dbData'),
    }
  })

  it('should verify token', () => {
    bestForYears(null, { ranking: 10 }, { db, token: 'test' })

    expect(verifyToken).toHaveBeenCalledWith('test')
  })

  it('should get the best movies for each year', async () => {
    await bestForYears(null, { ranking: 10 }, { db })

    expect(db.query.mock.calls[0][0]).toMatchSnapshot()
  })

  it('should return database data', async () => {
    const response = await bestForYears(null, { ranking: 10 }, { db })

    expect(response).toEqual('dbData')
  })

  it('should handle errors', async () => {
    db.query.mockImplementation(() => Promise.reject('nope'))

    try {
      await bestForYears(null, { ranking: 10 }, { db })
    } catch (e) {
      expect(e.message).toMatch(/nope/)
    }
  })
})
