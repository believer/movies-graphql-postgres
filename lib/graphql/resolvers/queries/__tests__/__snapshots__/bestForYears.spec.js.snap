// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries/bestForYears should get the best movies for each year 1`] = `
"
      SELECT * FROM (
        SELECT 
          m.title,
          m.release_date,
          r.rating,
          row_number() OVER (PARTITION BY date_part('year', release_date) ORDER BY rating DESC) AS ranking
        FROM 
          movies AS m 
          INNER JOIN ratings AS r ON r.movie_id = m.id
          WHERE r.user_id = 2
      ) ratings
      WHERE ratings.ranking = 10 AND ratings.release_date IS NOT NULL
      ORDER BY release_date DESC
    "
`;
